core/src/main/java/jenkins/security/ResourceDomainRootAction.java

```diff
@@ -0,0 +1,307 @@
+            return null;
+        }
+
+        Token token = Token.decode(id);
+        if (token == null) {
+            rsp.sendError(404, ResourceDomainFilter.ERROR_RESPONSE);
+            return null;
+        }
+
+        String authenticationName = token.username;
+        String browserUrl = token.path;
+
+
+        if (token.timestamp.plus(VALID_FOR_MINUTES, MINUTES).isAfter(now()) && token.timestamp.isBefore(now())) {
+            return new InternalResourceRequest(browserUrl, authenticationName);
+        }
+
+        // too old, so redirect to the real file first
+        return new Redirection(browserUrl);
+    }
+
+    private static class Redirection {
+        private final String url;
+
+        private Redirection(String url) {
+            this.url = url;
+        }
+
+        public void doDynamic(StaplerRequest req, StaplerResponse rsp) throws IOException {
+            String restOfPath = req.getRestOfPath();
+
+            String url = Jenkins.get().getRootUrl() + this.url + restOfPath;
+            rsp.sendRedirect(302, url);
+        }
+    }
+
+    public String getRedirectUrl(@Nonnull Token token, @Nonnull String restOfPath) {
+        String resourceRootUrl = getResourceRootUrl();
+        if (!restOfPath.startsWith("/")) {
+            // Unsure whether this can happen -- just be safe here
+            restOfPath = "/" + restOfPath;
+        }
+        return resourceRootUrl + getUrlName() + "/" + token.encode() + restOfPath;
+    }
+
+    private static String getResourceRootUrl() {
+        return ResourceDomainConfiguration.get().getUrl();
+    }
+
+    /**
+     * Called from {@link DirectoryBrowserSupport#generateResponse(StaplerRequest, StaplerResponse, Object)} to obtain
+     * a token to use when rendering a response.
+     *
+     * @param dbs the {@link DirectoryBrowserSupport} instance requesting the token
+     * @param req the current request
+     * @return a token that can be used to redirect users to the {@link ResourceDomainRootAction}.
+     */
+    @CheckForNull
+    public Token getToken(@Nonnull DirectoryBrowserSupport dbs, @Nonnull StaplerRequest req) {
+        // This is the "restOfPath" of the DirectoryBrowserSupport, i.e. the directory/file/pattern "inside" the DBS.
+        final String dbsFile = req.getRestOfPath();
+
+        // Now get the 'restOfUrl' after the top-level ancestor (which is the Jenkins singleton).
+        // In other words, this is the complete URL after Jenkins handled the top-level request.
+        final String completeUrl = req.getAncestors().get(0).getRestOfUrl();
+
+        // And finally, remove the 'restOfPath' suffix from the complete URL, as that's the path from Jenkins to the DBS.
+        String dbsUrl = completeUrl.substring(0, completeUrl.length() - dbsFile.length());
+        LOGGER.fine(() -> "Determined DBS URL: " + dbsUrl + " from restOfUrl: " + completeUrl + " and restOfPath: " + dbsFile);
```

Comment:
For reference, this block caused https://issues.jenkins-ci.org/browse/JENKINS-59849